Source/NameServer/AddressBook/AddressBook.cpp

```diff
@@ -0,0 +1,1367 @@
+        TRec->ExtCnt++;
+        DRec = &TRec->Externals.back();
+
+        TRec->AddrMap.insert(AddrMap_t::value_type(DRec->Address, DRec));
+        MapDevice(TRec, DRec);          // Add dev to other maps & lists.
+
+    } else if (DevRec.RecordType == Supervisor) {
+        // Add a Supervisor
+        // DebugPrint("Supervisor device being added\n");
+        CheckVector(TRec, TRec->Supervisors);  // Sanity check for reallocation.
+
+        TRec->Supervisors.push_back(DevRec);
+        TRec->SupCnt++;
+        DRec = &TRec->Supervisors.back();
+
+        TRec->AddrMap.insert(AddrMap_t::value_type(DRec->Address, DRec));
+
+        // Add the supervisor to the correct DeviceType list.
+        TRec->DevTypes[DRec->DeviceType].second.push_back(DRec);
+
+        SupMap_t::iterator SSearch = TRec->SupMap.find(DRec->Address);
+        if (SSearch == TRec->SupMap.end())
+        {   //Supervisor is NOT in the map yet.
+            TRec->SupMap.insert(SupMap_t::value_type(DRec->Address,
+                                    RecordVect_t()));
+        }
+    }
+    // added 18 July 2019 ADR - if Map and Link were previously invalidated
+    // rebuild everything
+    //
+    // Reverted by GMB as this potentially means heafty rebuilding
+    // when futzing with the data structure - the correct course of
+    // action is to check the task validity after adding devices and 
+    // rebuilding the link and map then if required.
+    //
+    //unsigned err;
+    //if (!TRec->MapValid && ((err = BuildMaps(TRec->Name)) != SUCCESS)) return err;
+    //if (!TRec->LinkValid && ((err = BuildLink(TRec->Name)) != SUCCESS)) return err;
+    
+    return SUCCESS;
+}
+
+
+/*==============================================================================
+ * UpdateDevice:    Update a device with a new address and supervisor.
+ *                  This implicitly marks the Linkage as invalid.
+ *============================================================================*/
+unsigned AddressBook::UpdateDevice(std::string &TaskName, DeviceData_t &Data)
+{
+    TaskRecord_t *TRec = FindTask(TaskName);
+    if(TRec == PNULL)
+    {
+        return(ERR_TASK_NOT_FOUND);
+    }
+
+    // Check that the Device Name exists
+    NameMap_t::iterator DSearch = TRec->NameMap.find(Data.Name);
+    if (DSearch == TRec->NameMap.end()) {
+        return ERR_DEVICE_NOT_FOUND;
+    }
+    Record_t *DRec = DSearch->second;
+
+    TRec->LinkValid = false;              // Mark the Task linkage as dirty
+    DRec->Address = Data.Address;         // Update the Device Address
+    DRec->Supervisor = Data.Supervisor;   // Update the Device Supervisor
+    return SUCCESS;
+}
+
+/*==============================================================================
+ * GetDeviceCount:  Get the number of devices reserved in the named task
+ *============================================================================*/
+long AddressBook::GetDeviceCount(std::string &TaskName)
```

Comment:
Returning a `long` here (and for some similar methods) is awkward, but I understand that, since `ERR_BAD_COUNT` is `-1`, you need to allow for negative numbers. The alternative would be to make `ERR_BAD_COUNT` equal to `ULONG_MAX` (from `<climits>`), and check against that. This means you don't need to worry about casting, making the code more readable.

Reply:
The only issue I can see with this is that then all of the returns have to be `unsigned long`

Reply:
True, but you wouldn't expect "GetDeviceCount" to return anything less than zero.

Reply:
I guess I am just not keen on ambiguous returns when changing the bit length of the return.